Links component refactor

wackerow opened the pull request:
- Creates a new component `Links` (Did not work right when I named it `Link`... not sure why)
- Accepts a prop `to` (string or object), or `href` (string, overrides `to`), and checks if it is an external link (starts with `http` or `mailto:`).
- If so, styles it with an arrow, and opens it in a new tab
- Has option to `hideArrow` (boolean prop, default false)
- Optional props for additional classes and aria labels

wackerow committed:
Refactors a `Links` component to replace `a` and `router-link`

Automatically styles link depending on if it's internal or external.

wackerow committed:
fix router-link prop name

corwintines commented on vue-app/src/components/RoundInformation.vue:
```diff
@@ -207,20 +207,20 @@
-                </a>
-                <a
+                </links>
+                <links
```
I think this is causing some errors because there is no "to", therefore its being sent as undefined and causing that error in the router-link

wackerow committed:
reverts to `spans` styled as a link

uses @click handler, not `to` attribute

wackerow committed:
allows passage of objects to Links component

passes to <router-link> unchanged for links with params

wackerow committed:
temporarily remove buggy Tooltip component

This component is being deprecated in favor of v-tooltip component. Was causing a console warning on this component, and newer version with v-tooltip directive should take it's place during merge conflict resolution

wackerow committed:
Merge branch 'develop' into links

corwintines reviewed (approved):
Nice 🚀

pettinarip commented on vue-app/src/components/Links.vue:
```diff
@@ -0,0 +1,72 @@
+    rel="noopener"
+  >
+    <slot />
+  </a>
```
Using object syntax to define dynamic classes https://vuejs.org/v2/guide/class-and-style.html#Object-Syntax I think we can make this simpler.
```js
<a
  v-if="isExternal"
  :class="{ 'external-link': !hideArrow, [className]: !!className }"
>
  <slot />
</a>
```

pettinarip replied:
[pending thought]
I'm wondering whether there is a way in Vue to avoid using a different name (`className`) to pass a custom class. I would ideally use `class` from the consumer.
```js
// right now
<links :className="....">

// what I would really like
<links :class="....">
```

wackerow replied:
More I think about it, do we even need to pass a prop for this at all? Couldn't we just assign a class to <links> wherever it is used and customize styling as needed in the parent component?

ie. `BackToProjects.vue`
We assign classes to the <links> component that is embedded. The CSS is declared within this file, and is used to custom style the resulting link, without passing 'className' prop

Testing locally, I tried overriding classes for external links using this method as well and seems to be working. I think this prop can be removed entirely

pettinarip replied:
Ha! read this https://vuejs.org/v2/guide/components-props.html#Replacing-Merging-with-Existing-Attributes
It will merge them (only for `class` and `style` attrs), cool!

So, yes, we are good if you remove that `className`. Any attr defined from the consumer perspective will be automatically spread in the root element of the component.

pettinarip commented on vue-app/src/components/MobileTabs.vue:
```diff
@@ -1,6 +1,6 @@
+    <links
       v-for="({ icon, title, to }, idx) of tabs"
       :key="idx"
       :class="{
```
This should be `className` now, right?

pettinarip commented on vue-app/src/components/BrightIdModal.vue:
```diff
@@ -57,13 +57,15 @@ import {
   components: {
     Transaction,
     Loader,
+    Links,
```
Are we using the Links component? or do we need some cleanup here?

wackerow replied:
At least on this branch, line 16 uses this component.

pettinarip replied:
Oh yes, I clicked on the wrong file. It was supposed to be `vue-app/src/components/CallToActionCard.vue`

pettinarip commented on vue-app/src/components/BrightIdWidget.vue:
```diff
@@ -70,10 +70,11 @@ import Component from 'vue-class-component'
 // TODO clean up this component
 @Component({
-  components: { Tooltip, IconStatus },
+  components: { Tooltip, IconStatus, Links },
```
Here too. Do we want to cleanup this?

wackerow replied:
👍🏻 Will remove this

pettinarip commented on vue-app/src/views/Profile.vue:
```diff
@@ -71,13 +71,14 @@ import BalanceItem from '@/components/BalanceItem.vue'
 
 @Component({
-  components: { BalanceItem, BrightIdWidget, IconStatus, CopyButton },
+  components: { BalanceItem, BrightIdWidget, IconStatus, CopyButton, Links },
```
Cleanup?

wackerow replied:
missed this, just patched 👍🏻

pettinarip reviewed (changes requested):
Good job refactoring and replacing all the links we have.

My main concern is around the classes. How we pass dynamic classes being defined at the consumer level down to the dynamic classes we have in the `Links` component itself?

```js
// consumer level
<links :class="{ dynamic: condition }">

// links component
<a :class={ componentClass: condition, ...classesInheritFromConsumer }>
// how do we spread those inherit dynamic classes here?
```

wackerow committed:
Links cleanup

- Consolidate `isExternal` class logic
- Remove unnecessary  `className` prop from Links  (can override styling from parent component with traditional `class` attribute)
- Made ↗ arrow white

wackerow committed:
Links refactor cleanup

pettinarip reviewed (approved):
LGTM! 💯

wackerow committed:
removed unused Links import

wackerow committed:
Merge branch 'develop' into links

samajammin reviewed (changes requested):
Looks solid! A few thoughts:

1. I think the external link arrows should be the same color as the link:

![clr fund - Can I add this 2021-08-16 at 1 32 18 PM](https://user-images.githubusercontent.com/8097623/129625989-0ec66f64-8618-47fb-893d-8f4c009b989a.jpg)

2. When links don't have text (e.g. these icon links) we shouldn't display the arrow:
![clr fund - Can I add this 2021-08-16 at 1 32 10 PM](https://user-images.githubusercontent.com/8097623/129626010-8a59d917-40cb-4c27-9bd6-5cdcb563cb49.jpg)

3. I do agree with @pettinarip's point on passing dynamic classes - did we solve for this? Are there any places we actually need this? If not, maybe this could be done in a separate PR...

wackerow committed:
hide arrow on icon-only links

revert to default color on &:after link arrow

wackerow commented:
![image](https://user-images.githubusercontent.com/54227730/129758920-b2ccf1b4-c740-475e-8272-8dc38cc67aaa.png)
 @samajammin This should fix the issues you commented on. Arrow color should match surrounding text, and added `:hideArrow="true"` to icon-only external links. Will clear up merge conflicts.

wackerow committed:
Merge branch 'develop' into links
